core/arch/arm/pta/device.c

```diff
@@ -0,0 +1,98 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (C) 2019, Linaro Limited
+ */
+
+/*
+ * This pseudo TA is used by Linux TEE driver to fetch pseudo/early TA's UUIDs
+ * which can act as TEE bus devices.
+ */
+
+#include <kernel/pseudo_ta.h>
+#include <kernel/tee_ta_manager.h>
+#include <tee/uuid.h>
+#include <pta_device.h>
+#include <string.h>
+#include <user_ta_header.h>
+
+#define PTA_NAME "device.pta"
+
+static uint32_t num_devices;
+
+static TEE_Result get_num_devices(uint32_t types,
+			       TEE_Param params[TEE_NUM_PARAMS])
+{
+	const struct pseudo_ta_head *ta;
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_VALUE_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	num_devices = 0;
+
+	SCATTERED_ARRAY_FOREACH(ta, pseudo_tas, struct pseudo_ta_head) {
+		if (ta->flags & TA_FLAG_KERNEL_DEVICE)
+			num_devices++;
+	}
+
+	params[0].value.a = num_devices;
+
+	return TEE_SUCCESS;
+}
+
+static TEE_Result get_devices(uint32_t types,
+			      TEE_Param params[TEE_NUM_PARAMS])
+{
+	const struct pseudo_ta_head *ta;
+	TEE_UUID *device_uuid = NULL;
+	uint8_t uuid_octet[sizeof(TEE_UUID)];
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE)) {
+		EMSG("bad parameters types: 0x%" PRIx32, types);
+		return TEE_ERROR_BAD_PARAMETERS;
+	}
+
+	if (params[0].memref.size < (num_devices * sizeof(TEE_UUID))) {
```

Comment:
Instead of this test you could check inside the loop below that there's still room in the target buffer. That would save us a global variable.

Reply:
Ok will put a check inside loop.